### DISCUSSION TITLE
Features: Deployment Pipe & Redirects

### DISCUSSION THREAD

**Discussion by casparjones** [caprover/caprover#1271] [Ideas]
I want to avoid being a pain, but I have a few ideas and wanted to hear what you think.
We like to use Heroku in the company and also on your homepage Heroku is compared with Caprover.
A very popular feature is the possibility to set up deployment pipes in Heroku. 
App1 = Staging -> App2 = Production with one click, you can copy App1 to App2 after testing.
Would you consider such a feature for Caprover as well?

Another feature might be a possibility to set up a redirect to the "www.*" domain without editing the default nginx configuration. This would probably be possible with it too, but with an option, it would be nice and easy.

**Comment by githubsaturn** [comment]
> App1 = Staging -> App2 = Production with one click, you can copy App1 to App2 after testing.
Would you consider such a feature for Caprover as well?

Heroku is only addressing the simple case of very simple apps, no persistent volumes, no dependency on the host etc. If that's what you like, all you need to do is this:
- Create a staging and a production app on your instance `staging-my-app` and `prod-my-app`
- When developing and pushing, only push to `staging-my-app`
- Once you're happy with the result, simply just deploy the image on the prod app. You can do this directly on the UI (Method 6) or using CLI `--imageName` 

Does this address your usecase?

![image](https://user-images.githubusercontent.com/10582082/143144933-a702ed1a-21e8-4846-aace-36b9b8163ecb.png)



> Another feature might be a possibility to set up a redirect to the "www.*" domain without editing the default nginx configuration. This would probably be possible with it too, but with an option, it would be nice and easy.

It's already a requested feature: https://github.com/caprover/caprover/issues/314
It'll come at some point, the UX just needs to be flushed out.

**Comment by casparjones** [comment]
I have not thought about this possibility and thought that deployment method No. 6 only works with images from [dockerhub](https://hub.docker.com/). I just tried it on my DEV server and got the error message:
Error: (HTTP code 404) unexpected - pull access denied for img-captain-xxx, the repository does not exist or may require 'docker login': denied: requested access to the resource is denied

That the feature for the redirection in planning is pleased me very much. :)
I gave a small presentation to my colleagues for Caprover this week, and they were also very excited, but that's when the question about the deployment pipe came up.

**Reply by githubsaturn** [reply]
> Error: (HTTP code 404) unexpected - pull access denied for img-captain-xxx, the repository does not exist or may require 'docker login': denied: requested access to the resource is denied

This is a shortcoming that needs to be addressed. Feel free to create an issue for this. "Local images do not need to be pulled"
As a workaround you can try Method 4, and just have have one line:
```Dockerfile
FROM img-captain-xxx
```

**Reply by casparjones** [reply]
Wow, so cool, this works great with method 4 and FROM.... 
Thanks for Caprover and all the work you put in here!
